File: doc/R/opts.R
#!/usr/local/bin/Rscript
#
#    opts.R
#    for easy use of getopt(1).
#    $Id: opts.R,v 1.1 2024/04/17 14:25:23 daichi Exp $
#
library("getopt")

getopts <- function (spec, usage, n, ...)
{
    res <- list()
    tokens <- commandArgs (T)
    
    if (length(tokens) < n) {
        usage ()
    }

    if (substr(tokens[1], 1, 1) == "-") {
        res$args <- tail (tokens, n)
        res$opts <- getopt (matrix (spec, byrow=T, ncol=4),
                            head (tokens, -n), ...)
    } else {
        res$args <- head (tokens, n)
        res$opts <- getopt (matrix (spec, byrow=T, ncol=4),
                            tail (tokens, -n), ...)
    }

    # fill unused options by NA
    names <- spec [seq(1,length(spec),by=4)]
    for (name in names) {
        if (is.null(res$opts[[name]])) {
            res$opts[[name]] <- NA
        }
        res$opts$ARGS <- NULL
    }

    res
}

main <- function ()
{
    usage <- function ()
    {
        cat ('usage: % opts.R --iter=iters --output=file train model\n')
        quit ()
    }

    opts <- getopts ( c(
        "alpha",  "a", 1, "double",
        "iters",  "N", 2, "integer",
        "output", "o", 2, "character",
        "help",   "h", 0, "logical"
        ), usage, 2)

    if (!is.na(opts$opts$help)) {
        usage ()
    }

    print (opts$args)
    print (opts$opts)

}


if (!interactive() & sys.nframe() == 0) {
    main ()
}

File: doc/bin/opts.py
#!/usr/local/bin/python
#
#    opts.py
#    $Id: opts.py,v 1.4 2019/03/17 01:21:37 daichi Exp $
#
import re
import sys
import getopt

def optchar (s):
    if s[-1] == '=':
        s = s[0:-1]
    if s[0:2] == '--':
        return s[2:]
    elif s[0] == '-':
        return s[1:]
    else:
        return s

def getopts (directive):
    dic   = {}
    chars = []
    longs = []
    short = ''
    # parse directives
    for item in directive:
        # option key
        [char,long] = item.split('|')
        if re.search(r'=$', long):
            short += ('%s:' % char)
        else:
            short += char
        chars.append (char)
        longs.append (long)
    # analyze argv
    try:
        opts,args = getopt.getopt (sys.argv[1:], short, longs)
    except getopt.GetoptError as err:
        usage ()
    # register dictionary
    for o,a in opts:
        oo = optchar (o)
        for char,long in zip (chars, longs):
            long = (long[0:-1] if long[-1] == '=' else long)
            if oo in (char, long):
                dic[char] = (True if a == '' else a)
                break
    return dic, args

def getopts1 (optarray):
    chars = []
    longs = []
    short = ''
    # parse directives
    for item in optarray:
        [char,long] = item.split('|')
        if re.search(r'=$', long):
            short += ('%s:' % char)
        else:
            short += char
        chars.append (char)
        longs.append (long)
    # analyze argv
    try:
        opts,args = getopt.getopt (sys.argv[1:], short, longs)
    except getopt.GetoptError as err:
        usage ()
    # create dictionary
    dic = {}
    for o,a in opts:
        oo = optchar (o)
        for char,long in zip (chars, longs):
            long = (long[0:-1] if long[-1] == '=' else long)
            if oo in (char, long):
                dic[char] = (None if a == '' else a)
                break
    return dic,args

# for test
# 
# def main ():
#     # opts,args = getopts ({"L|latents=":2, "f|foo=":"bar", "h|help":None})
#     opts,args = getopts (["L|latents=", "f|foo=", "h|help"])
#     print (opts)
#     print (args)
#     if 'h' in opts:
#         print ('HELP!')
# 
# def usage ():
#     print ('usage: opts.py [-L latents] [-f arg] [-h] train model')
#     sys.exit (0)
# 
# if __name__ == "__main__":
#     main ()
